### DISCUSSION TITLE
exporting .h5 file has several thousand track identities

### DISCUSSION THREAD

**Discussion by mollywestbrook** [talmolab/sleap#1310] [Help!] [answered]
Hello! I was able to get predictions for a 30 minute video (videos are typically ~45k frames long) with 5 animals. Skeleton has 8 nodes. I was able to correct the video's tracklets to correspond to each animal, and the first 5 track identities correspond to those animals, but when I go to export my .h5, I get this error pop-up in the terminal: 

`numpy.core._exceptions.MemoryError: Unable to allocate 136. GiB for an array with shape (44983, 8, 2, 25420) and data type float64`

There are a number of other messages in the terminal, but I believe this line is the problem. From my understanding (and from issue #1238) sleap is not supposed to be exporting empty tracks, which I assume the other 25000+ track identities are. Is my model just not good enough to be tracking at this point, or is there a way for me to get rid of these unnecessary track labels quickly? Thank you for any help! Using v.1.2.9

**Comment by roomrys** [comment, marked as answer]
Hi @mollywestbrook,

That is a lot of tracks. SLEAP will remove empty tracks before returning the final analysis file; however, SLEAP still initializes the occupancy matrix with the total number of tracks in the project - which seems to be the cause of error in this case.

Can you try deleting the empty tracks and saving before doing the export:
![image](https://github.com/talmolab/sleap/assets/38435167/ff522469-1c11-4d4f-b902-b43ffdde2604)
> This feature is only available in 1.3.0+

The code to do this through the API (which should work in 1.2.9) would be:
```python
"""Modified from Labels.remove_unused_tracks (v1.3.0)"""
import sleap


labels = sleap.load_file(r"path/to/slp")

# Check which tracks are used by instances
all_tracks = set(labels.tracks)
used_tracks = set()
for inst in labels.instances():
    used_tracks.add(inst.track)

# Remove set difference from tracks in Labels
tracks_to_remove = all_tracks - used_tracks
for track in tracks_to_remove:
    labels.tracks.remove(track)

# Save to a new filename (not the same filename!)
labels.save(r"path/to/new/slp")
```

The code to do this through the API for 1.3.0+ would be:
```python
import sleap


labels = sleap.load_file(r"path/to/slp")

# Remove unused tracks
labels.remove_unused_tracks()

# Save to a new filename (not the same filename!)
labels.save(r"path/to/new/slp")

```


Let us know how that goes.

Thanks,
Liezl

**Comment by mollywestbrook** [comment]
Works great Liezl, thank you for the quick response and fix! I went with upgrading the version of sleap I was using to 1.3.0 in a fresh environment and using the GUI, and the .h5 file looks great.
